### PR TITLE
[WIP] support distributed pk

### PR DESCRIPTION
**Pull request by hzliangbin**
Signed-off-by: Bin Liang <hzliangbin@foxmail.com>

ref: https://github.com/cectc/dbpack/issues/42

### CODE REVIEW

File: pkg/misc/uuid/generator.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package uuid
```
Comment (dk-lockdown) [inline pkg/misc/uuid/generator.go:17]: please add license header

File: pkg/misc/uuid/segment.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
+package uuid
```
Comment (dk-lockdown) [inline pkg/misc/uuid/segment.go:17]: same as above

File: pkg/misc/uuid/snowflake.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package uuid
```
Comment (dk-lockdown) [inline pkg/misc/uuid/snowflake.go:17]: same as above

File: pkg/misc/uuid/snowflake.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+func (w *SnowflakeWorker) NextID() (int64, error) {
+	ts := time.Now().UnixMilli()
+	if ts < w.lastTs {
+		offset := w.lastTs - ts
+		if offset > 5 {
+			return int64(0), fmt.Errorf("worker %d clock move back", w.workerID)
+		}
+		time.Sleep(time.Duration(offset<<1) * time.Millisecond)
+		ts = time.Now().UnixMilli()
```
Comment (dk-lockdown) [inline pkg/misc/uuid/snowflake.go:59]: please fix ci check failure

File: pkg/misc/uuid/store.go [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package uuid
+
+type Store interface {
+	List() ([]*Segment, error)
+	Get() (*Segment, error)
+	UpdateMaxID( ) (*Segment, error)
+	UpdateMaxIDWithStep(step int32) (*Segment, error)
+}
```
Comment (dk-lockdown) [inline pkg/misc/uuid/store.go]: please add a blank line at the end

File: pkg/misc/uuid/segment.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+func (w *SegmentWorker) PreloadBuffer(ctx context.Context) error {
+	for i := 0; i < MaxRetry; i++ {
+
+	}
+	return nil
+}
+
+func (w *SegmentWorker) HasID(id int64) bool {
+	return id != 0
+}
```
Comment (dk-lockdown) [inline pkg/misc/uuid/segment.go]: please add a blank line at the end

File: pkg/misc/uuid/segment.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+const (
+	MaxRetry = 3
+)
+
+type SegmentWorker struct {
+	Key        string
+	Step       int32
+	CurrentPos int32                  // current buffer pos
+	Buffer     []*Segment             // use two buffer, one buffer offers ID, another as preload
+	UpdateTime time.Time              //  mark update time
```
Comment (bohehe) [inline pkg/misc/uuid/segment.go]: extra space should be removed

File: pkg/misc/uuid/segment.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	mutex      sync.Mutex             //
```
Comment (bohehe) [inline pkg/misc/uuid/segment.go]: remove `//` or leave a comment after it

File: pkg/misc/uuid/segment.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+type SegmentRecord struct {
+	ID         int64  `json:"id" form:"id"` // pk
+	BizTag     string `json:"biz_tag" form:"biz_tag"`
+	MaxID      int64  `json:"max_id" form:"max_id"`
+	Step       int32  `json:"step" form:"step"`
+	UpdateTime int64  `json:"update_time" form:"update_time"`
+}
+
+func NewSegmentWorker(sr *SegmentRecord) (*SegmentWorker, error) {
```
Comment (bohehe) [inline pkg/misc/uuid/segment.go]: remove error if it always returns nil

File: pkg/misc/uuid/segment.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+		w.UpdateTime = time.Now()
+	}
+	return id, nil
+}
+
+func (w *SegmentWorker) HasSeq() bool {
+	if w.Buffer[w.CurrentPos].InitOk && w.Buffer[w.CurrentPos].Cursor < w.Buffer[w.CurrentPos].Max {
+		return true
+	}
+	return false
```
Comment (bohehe) [inline pkg/misc/uuid/segment.go]:
```suggestion
	return w.Buffer[w.CurrentPos].InitOk && w.Buffer[w.CurrentPos].Cursor < w.Buffer[w.CurrentPos].Max
```

**Review by bohehe — APPROVED** [review]

**Review by dk-lockdown — APPROVED** [review]
how about provide SegmentWorker sql scirpts

**Pull request merged**